source/util/bit_stream.h

```diff
@@ -0,0 +1,371 @@
+  // Note: "01" will be writen as 0x2, not 0x1. The string doesn't represent
+  // numbers but a stream of bits in the order they come from encoder.
+  virtual void WriteStream(const std::string& bits) {
+    WriteBits(StreamToBits(bits), bits.length());
+  }
+
+  // Writes lower |num_bits| in |bits| to the stream.
+  // |num_bits| must be no greater than 64.
+  template <size_t N>
+  void WriteBitset(const std::bitset<N>& bits, size_t num_bits = N) {
+    WriteBits(bits.to_ullong(), num_bits);
+  }
+
+  // Writes |val| in chunks of size |chunk_length| followed by a signal bit:
+  // 0 - no more chunks to follow
+  // 1 - more chunks to follow
+  // for example 255 is encoded into 1111 1 1111 0 for chunk length 4.
+  // The last chunk can be truncated and signal bit omitted, if the entire
+  // payload (for example 16 bit for uint16_t has already been written).
+  void WriteVariableWidthU64(uint64_t val, size_t chunk_length);
+  void WriteVariableWidthU32(uint32_t val, size_t chunk_length);
+  void WriteVariableWidthU16(uint16_t val, size_t chunk_length);
+  void WriteVariableWidthU8(uint8_t val, size_t chunk_length);
+  void WriteVariableWidthS64(
+      int64_t val, size_t chunk_length, size_t zigzag_exponent);
+  void WriteVariableWidthS32(
+      int32_t val, size_t chunk_length, size_t zigzag_exponent);
+  void WriteVariableWidthS16(
+      int16_t val, size_t chunk_length, size_t zigzag_exponent);
+  void WriteVariableWidthS8(
+      int8_t val, size_t chunk_length, size_t zigzag_exponent);
+
+  // Returns number of bits written.
+  virtual size_t GetNumBits() const = 0;
+
+  // Provides direct access to the buffer data if implemented.
+  virtual const uint8_t* GetData() const {
+    return nullptr;
+  }
+
+  // Returns buffer size in bytes.
+  size_t GetDataSizeBytes() const {
+    return NumBitsToNumWords<8>(GetNumBits());
+  }
+
+  // Generates and returns byte array containing written bits.
+  virtual std::vector<uint8_t> GetDataCopy() const = 0;
+
+  BitWriterInterface(const BitWriterInterface&) = delete;
+  BitWriterInterface& operator=(const BitWriterInterface&) = delete;
+};
+
+// This class is an implementation of BitWriterInterface, using
+// std::vector<uint64_t> to store written bits.
+class BitWriterWord64 : public BitWriterInterface {
+ public:
+  explicit BitWriterWord64(size_t reserve_bits = 64);
+
+  void WriteBits(uint64_t bits, size_t num_bits) override;
+
+  size_t GetNumBits() const override {
+    return end_;
+  }
+
+  const uint8_t* GetData() const override {
+    return reinterpret_cast<const uint8_t*>(buffer_.data());
+  }
+
+  std::vector<uint8_t> GetDataCopy() const override {
+    return std::vector<uint8_t>(GetData(), GetData() + GetDataSizeBytes());
+  }
+
+  // Returns written stream as std::string, padded with zeroes so that the
+  // length is a multiple of 64.
+  std::string GetStreamPadded64() const {
+    return BufferToStream(buffer_);
+  }
+
+ private:
+  std::vector<uint64_t> buffer_;
+  size_t end_;
```

Comment:
end_ member needs a comment.   It's no obvious what it is.  From GetNumBits it seems it's the number of bits written so far.